src/Temporalio/Worker/WorkflowCodecHelper.cs

```diff
@@ -0,0 +1,279 @@
+using System.Linq;
+using System.Threading.Tasks;
+using Google.Protobuf.Collections;
+using Temporalio.Api.Common.V1;
+using Temporalio.Bridge.Api.ActivityResult;
+using Temporalio.Bridge.Api.ChildWorkflow;
+using Temporalio.Bridge.Api.WorkflowActivation;
+using Temporalio.Bridge.Api.WorkflowCommands;
+using Temporalio.Bridge.Api.WorkflowCompletion;
+using Temporalio.Converters;
+
+namespace Temporalio.Worker
+{
+    /// <summary>
+    /// Utilities to encode/decode workflow protobufs.
+    /// </summary>
+    internal static class WorkflowCodecHelper
+    {
+        /// <summary>
+        /// Encode the completion.
+        /// </summary>
+        /// <param name="codec">Codec to use.</param>
+        /// <param name="comp">Completion to encode.</param>
+        /// <returns>Task for completion.</returns>
+        internal static async Task EncodeAsync(
+            IPayloadCodec codec, WorkflowActivationCompletion comp)
+        {
+            switch (comp.StatusCase)
+            {
+                case WorkflowActivationCompletion.StatusOneofCase.Failed:
+                    if (comp.Failed.Failure_ != null)
+                    {
+                        await codec.EncodeFailureAsync(comp.Failed.Failure_).ConfigureAwait(false);
+                    }
+                    break;
+                case WorkflowActivationCompletion.StatusOneofCase.Successful:
+                    foreach (var cmd in comp.Successful.Commands)
+                    {
+                        await EncodeAsync(codec, cmd).ConfigureAwait(false);
```

Comment:
I am not convinced there is a ton of value in parallelizing. In fact, I think there's more value in batching. Shall we change our codecs to submit these things in batches? Or do we accept code complication is not enough benefit to contort the code to try to do multiple of these things at once?

Reply:
Value is shorter task processing times and reduced chance of workflow task timeouts especially for remote codecs.
Not blocking the PR, we don't parallelize all invocations in other SDKs.

Reply:
Just thought it would be nice to consider since this is a fresh implementation.

Reply:
You would be much shorter processing times for remote codecs if you batched. Should we batch instead?

Reply:
I agree that batching is better approach than parallelizing for both remote and non remote codecs.

We should just make sure that a single batch correlates to a single "context" (workflow / activity) to simplify the user API for when they need their codecs to be context aware.

Reply:
For performance reasons, should we update SDKs to support batching? Should we update .NET to support batching? Or should we accept that performance benefit is negligible when doing separate requests.

Reply:
I don't know how negligible it is, I think there are definitely cases where it would matter.
In any case I don't think we want to take this on for all SDKs right now.
In the meantime, batching could easily be built on top of concurrency user side by waiting a tick before submitting any accumulated requests (similarly to how this is done in [dataloader](https://www.npmjs.com/package/dataloader)) .

I wouldn't necessarily change the API and wouldn't block this PR, I would just take this into consideration.
In TypeScript we try to run codecs concurrently as much as possible (but I admit that there's room for improvement).

Reply:
While attempting to parallelize this, two other reasons came up to potentially not do it: 1) this will aggregate exceptions, so you could throw 50 exceptions from here for 50 commands, and 2) This can explode remote/IO call count filling up the thread pool. Technically they can prevent this by debouncing, but it is not as clear.

If you have a codec that does CPU bound AES let's say, and I call this concurrently for all 50 commands, .NET will try to use as much of the thread pool as it has available. Scatter-gather makes less sense for codecs for potentially IO/CPU-bound tasks in multithreaded languages.